### PR TITLE
Add HUD-based summarize harness for repeated_poker

### PR DESCRIPTION
**Pull request by bovard**
Adds summarize_harness.py, a cost-reduced alternative to harness.py that replaces the full per-hand transcript (which grows quadratically over a 100-hand session) with a compact, deterministic opponent HUD:

- Standard tracker stats computed from the structured hand histories: VPIP, PFR, 3Bet, Fold-to-3Bet, CBet, Fold-to-CBet, WTSD, AF, plus net.
- A context-aware per-street read of the opponent's tendencies on the street the current decision is on.
- Full renders of only the last 2 hands for concrete recency.
- A tightened prompt template (~690 -> ~260 tokens) that preserves the load-bearing Final Answer format and bet-sizing convention so the existing response parser is unaffected.

Parsing, legal-move extraction, and rethink handling are imported verbatim from harness.py; only the readable-state construction differs. Measured ~9x reduction in input tokens over a full 100-hand session (10x+ in the later hands where history would otherwise dominate).

### CODE REVIEW

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+#   - the literal "Final Answer:" tag (RuleBasedMoveParser splits on it),
+#   - the five action words fold/check/call/bet/raise,
+#   - the "total for the round" bet-sizing convention (soft parser assumes it).
+# The GTO prose, the concept checklist, and the duplicated examples are dropped.
+# ---------------------------------------------------------------------------
+
+REPEATED_POKER = """
+You are a world-class Heads-Up No-Limit Texas Hold'em poker AI. Play to maximize
+EV: use GTO as a baseline and deviate to exploit the opponent's tendencies (see
+the HUD and per-street reads below). Consider ranges, board texture, position,
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py]: **Major:** The preamble points the model at "per-street reads", but that feature was removed in `b21ad180` (see `del cur_parse_state` at line 312) and nothing renders it. This fires on 100% of prompts — suggest dropping `and per-street reads`.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+    sections: list[str] = [f"You are Player{cur}."]
+    sections.append(_render_standing(state_dict, cur))
+
+    if stats.hands > 0:
+        sections.append(_render_opponent_model(stats, opp))
+
+        recent_idx = deep_hand_indices[-_NUM_RECENT_HANDS:]
+        if recent_idx:
+            rendered_recent = [_render_past_hand(acpc_hhs[i], (i % 2) + 1, cfg) for i in recent_idx]
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py]: **Major — hidden info leak:** `_render_past_hand` renders with `observer_id=None`, which prints *both* players' hole cards, and `_reached_turn_or_showdown` also selects hands that ended in a turn/river fold. I confirmed a generated prompt showing `Dealt to Player1 [4c 6h]` for a hand Player1 folded on the river — cards that were never revealed in-game; suggest passing `observer_id=f"Player{cur}"` unless the hand actually reached showdown.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+    # Accumulate the opponent model over all completed hands. Preflop and flop
+    # tendencies are captured by the HUD; only hands that reached the turn or a
+    # showdown are worth replaying in full (later-street lines and revealed
+    # holdings can't be summarized statistically).
+    stats = _OpponentStats()
+    acpc_hhs = list(pyspiel_state.acpc_hand_histories())
+    deep_hand_indices: list[int] = []
+    for i, acpc_hh in enumerate(acpc_hhs):
+        button_index = (i % 2) + 1
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py]: **Medium:** Re-deriving the button matches the engine's `dealer` under the default config, but with `rotate_dealer=False` (settable via `openSpielGameParameters`) it is wrong on half the hands, which mis-seats SB/BB and corrupts every positional HUD stat. `state_dict` is already parsed at line 293, so reading `state_dict["dealer"]` is a one-line fix.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+
+def generate_prompt(
+    observation: Mapping[str, Any],
+    move_history: list[str],
+    previous_response: str | None = None,
+    previous_action: str | None = None,
+) -> str:
+    """Build the LLM prompt with the compact opponent model."""
+    del move_history, previous_action  # not used in repeated_poker prompts
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py]: **Medium:** `previous_action` is dropped here, and the inherited `POKER_RETHINK` always says "a legal action could not be parsed" even when the move parsed fine and was merely illegal. `harness.py` is excused by GameArena byte-parity, but this template is already a rewrite — worth branching the rethink suffix on `previous_action`.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+def _pct(pair: list[int]) -> str:
+    num, den = pair
+    if den <= 0:
+        return "n/a"
+    return f"{round(100 * num / den)}%"
+
+
+def _af(stats: _OpponentStats) -> str:
+    if stats.af_calls == 0:
+        return "inf" if stats.af_aggressive > 0 else "n/a"
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py]: **Medium:** AF is the only stat rendered without its `(made/opportunities)` denominator — a real 9-hand prompt showed a bare `AF inf`, which reads as a maximally aggressive opponent but may be a single bet. Suggest something like `AF inf (3b+r/0c)` so it can be weighted like the rest of the HUD.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+        standing = f"AHEAD by {cur_net} chips"
+    elif cur_net < 0:
+        standing = f"BEHIND by {-cur_net} chips"
+    else:
+        standing = "EVEN"
+    if big_blind:
+        standing += f" ({cur_net / big_blind:+.1f} BB)"
+    return (
+        f"=== Standing (scored on cumulative chip profit) ===\n"
+        f"{standing}, hand {hand_number + 1}/{max_num_hands} ({hands_left} left)."
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py]: nit: `hand 1/20 (20 left)` is self-consistent as "including current", but reads as "20 more after this one". It's also redundant with `hand X/Y` — probably simplest to drop the parenthetical.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+    return f"{stats.af_aggressive / stats.af_calls:.1f}"
+
+
+def _reached_turn_or_showdown(hand: hh_utils.Hand) -> bool:
+    """True if the hand went past the flop -- i.e. a turn card was dealt (so
+    there was turn/river action) or it ended in a showdown. These are the hands
+    where full replay adds information the HUD can't capture; hands decided
+    preflop or on the flop are already summarized by the stats."""
+    # community is [flop(3), turn(1), river(1)] with later streets possibly empty.
+    turn_dealt = len(hand.community) >= 2 and len(hand.community[1]) >= 1
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py:275]: nit: All-in preflop run-outs satisfy `turn_dealt` but contain no postflop action, so they burn one of the two recency slots on ~400 chars the HUD already covers. Consider additionally requiring at least one postflop action event.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+                response = next(
+                    (e for e in flop if e.actor == opp and e.kind in _FACING_KINDS),
+                    None,
+                )
+                if response is not None and response.kind is ActionKind.FOLD:
+                    stats.fold_to_cbet[0] += 1
+
+    # --- Postflop aggression factor ---
+    for e in events:
+        if e.actor != opp or e.street not in _POSTFLOP_STREETS:
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py:242]: nit: AF lumps flop/turn/river together and there are no turn- or river-specific stats, so later-street reads rest entirely on the ≤2 rendered hands. Worth calling out in the module docstring, which currently only claims preflop and flop are fully captured.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+        # Aggression factor components (postflop).
+        self.af_aggressive = 0
+        self.af_calls = 0
+
+
+def _accumulate_hand(stats: _OpponentStats, hand: hh_utils.Hand, opp: int) -> None:
+    """Fold one parsed past hand's opponent actions into the HUD counters."""
+    stats.hands += 1
+    if opp < len(hand.profits):
+        stats.net += hand.profits[opp]
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py]: nit: `stats.net` is accumulated here but never rendered by `_render_opponent_model`. Either surface it in the HUD or drop the field.

**Review by lipovetz — COMMENTED** [review]
These comments are the output of a `review-harness` run over the new harness — static review against the live engine, no replay archive.

The headline claims held up: I measured **8.9×** fewer prompt chars over a full 100-hand session (15.6× by hands 90–99), 418 decision nodes rendered with zero exceptions, and all eight HUD stats validated against hand-crafted ACPC histories. The notes below are what didn't check out.

File: kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+        else:
+            aggressor_cbet = any(e.actor == pf_aggressor and e.kind is ActionKind.BET for e in flop)
+            if aggressor_cbet:
+                stats.fold_to_cbet[1] += 1
+                response = next(
+                    (e for e in flop if e.actor == opp and e.kind in _FACING_KINDS),
+                    None,
+                )
+                if response is not None and response.kind is ActionKind.FOLD:
+                    stats.fold_to_cbet[0] += 1
```
Comment (lipovetz) [inline kaggle_environments/envs/open_spiel_env/games/repeated_poker/summarize_harness.py:238]: **HUD c-bet and WTSD are corrupted by all-in-preflop runouts.**

`_accumulate_hand` counts a hand toward the CB denominator and the WTSD numerator whenever the flop was dealt — including hands where both players were all-in preflop and the board simply ran out. The opponent had no opportunity to act on those streets, so the HUD scores them as "declined to c-bet" and "went to showdown."

Measured across three independent 90-hand sessions (`repeated_poker` defaults, random agents):

| seed | CB reported | CB excl. runouts | WTSD reported | WTSD excl. runouts |
|---|---|---|---|---|
| 1 | 21% (6/28) | 30% (6/20) | 51% (22/43) | 30% (9/30) |
| 2 | 23% (5/22) | 33% (5/15) | 62% (28/45) | 39% (11/28) |
| 3 | 25% (5/20) | 36% (5/14) | 60% (24/40) | 48% (15/31) |

CB is understated by 9–11 points and WTSD overstated by 12–23 points in every seed. The combined read — "passive c-bettor who shows down a lot" — is close to the inverse of the truth, which is exactly the read a model would exploit by over-bluffing and value-betting thin.

Worth noting `_reached_turn_or_showdown` already screens out exactly these hands so they "don't waste a recency slot"; the same filter just isn't applied to the stats.

Suggested fix — compute `has_postflop_action` once and gate both blocks on it:

```python
has_postflop_action = any(e.street in _POSTFLOP_STREETS for e in events)
if saw_flop and has_postflop_action:
    stats.wtsd[1] += 1
    if went_to_showdown:
        stats.wtsd[0] += 1
```

...and likewise gate the CBet / Fold-to-CBet block below on `has_postflop_action`. VPIP/PFR/3B/F3B are unaffected — those are preflop, where the opponent always had a decision.

(Magnitudes above come from random agents, which shove more often than LLMs, so real-session contamination is likely smaller. The mechanism holds regardless, and the hydrated example prompt showed near-shove preflop sizing is common in real sessions too.)

**Review by lipovetz — APPROVED** [review]
Ideally John should look, but since he's out this week feel free to check in for experimentation.
